### PR TITLE
New: no-dupe-args rule (fixes #1880)

### LINKED ISSUE #1880: Rule suggestion : Arguments name should be unique
Hello.

I'd like to get yout input about this rule from SonarQube. Would that be interesting to get this rule in ESLint ?

> Arguments name should be unique.
Function arguments should all have different names to prevent any ambiguity. Indeed, if arguments have the same name, the last duplicated argument hides all previous arguments with the same name (those previous arguments remain available through arguments[i], so they're not completely inaccessible. This hiding makes no sense, clearly reduces understandability and maintanability, and obviously can be error prone. Furthermore, in JavaScript strict mode, declaring arguments with the same name produces an error. 

Thanks.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8961340-rule-suggestion-arguments-name-should-be-unique?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by xjamundx**
Here's a simple solution to the `no-dupe-args` rule. It's the simplest way I could think to solve the problem. I'd love some feedback of fancier solutions :)

### CODE REVIEW

File: docs/rules/README.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ The following rules point out areas where you might have made mistakes.
 * [no-control-regex](no-control-regex.md) - disallow control characters in regular expressions
 * [no-debugger](no-debugger.md) - disallow use of `debugger`
 * [no-dupe-keys](no-dupe-keys.md) - disallow duplicate keys when creating object literals
+* [no-dupe-args](no-dupe-args.md) - disallow duplicate arguments  in functions
```
Comment (xjamundx) [inline docs/rules/README.md]: multiple spaces...

File: lib/rules/no-dupe-args.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+     * @returns {void}
+     * @private
+     */
+    function checkParams(node) {
+        var currentParam = "";
+        var hasDupArgs = node.params && node.params.map(function(param) {
+            return param.name;
+        }).some(function(param, i, params) {
+            currentParam = param;
+            return params.indexOf(param) !== params.lastIndexOf(param);
```
Comment (xjamundx) [inline lib/rules/no-dupe-args.js]: `return i !== params.lastIndexOf(param);` might also work

File: lib/rules/no-dupe-args.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        });
+
+        if (hasDupArgs) {
+            context.report(node, "Duplicate param '{{key}}'.", { key: currentParam});
```
Comment (lo1tuma) [inline lib/rules/no-dupe-args.js]: You could use the location of `currentParam` here.

**Comment by xjamundx** [comment]
We also aren't worrying about multiple duplicates. `function (a, a, b, b) {}` will only warn once. Assume that's wrong as well!

File: docs/rules/no-dupe-args.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# No duplicate arguments (no-dupe-args)
+
+If you have a function that takes many arguments you may sometimes use the same argument name more than once.
```
Comment (nzakas) [inline docs/rules/no-dupe-args.md]: Useful to mention that duplicate arguments are flagged in strict mode right up top here.

Comment (xjamundx) [inline docs/rules/no-dupe-args.md]: Good call

File: lib/rules/no-dupe-args.js (hunk 3 above)
Comment (xjamundx) [inline lib/rules/no-dupe-args.js]: I think it's usually fine just to know the line number and the param name (unless you have soooo many params)

**Comment by xjamundx** [comment]
Fixed a few minor things and added 2 more tests.

File: lib/rules/no-dupe-args.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/**
+ * @fileoverview Rule to flag duplicate arguments
+ * @author Jamund Ferguson
+ */
```
Comment (nzakas) [inline lib/rules/no-dupe-args.js]: Don't forget your copyright. :)

File: tests/lib/rules/no-dupe-args.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/**
+ * @fileoverview Tests for no-dup-args
```
Comment (nzakas) [inline tests/lib/rules/no-dupe-args.js]: `dupe`

File: docs/rules/README.md [hunk 7]
```diff
@@ -13,6 +13,7 @@ The following rules point out areas where you might have made mistakes.
 * [no-control-regex](no-control-regex.md) - disallow control characters in regular expressions
 * [no-debugger](no-debugger.md) - disallow use of `debugger`
 * [no-dupe-keys](no-dupe-keys.md) - disallow duplicate keys when creating object literals
+* [no-dupe-args](no-dupe-args.md) - disallow duplicate arguments in functions
```
Comment (nzakas) [inline docs/rules/README.md]: Move up one line (alphabetical order)

**Comment by xjamundx** [comment]
Added the copyrights and a few minor things...

**Comment by nzakas** [comment]
Lgtm.

**Pull request merged**